Makefile

```diff
@@ -1,14 +1,29 @@
-GOFLAGS :=
-IMAGE_REPOSITORY_NAME ?= openshift
+.PHONY: all
 
-build:
-	go build -mod=vendor $(GOFLAGS) .
-.PHONY: build
+# Include the library makefile
+include $(addprefix ./vendor/github.com/openshift/library-go/alpha-build-machinery/make/, \
+    golang.mk \
+    targets/openshift/deps-gomod.mk \
+    targets/openshift/images.mk \
+)
 
-images:
-	imagebuilder -f Dockerfile -t $(IMAGE_REPOSITORY_NAME)/oauth-proxy .
-.PHONY: images
+IMAGE_REGISTRY?=registry.svc.ci.openshift.org
+
+# This will call a macro called "build-image" which will generate image specific targets based on the parameters:
+# $0 - macro name
+# $1 - target name
+# $2 - image ref
+# $3 - Dockerfile path
+# $4 - context directory for image build
+# It will generate target "image-$(1)" for building the image and binding it as a prerequisite to target "images".
+$(call build-image,ocp-oauth-proxy,$(IMAGE_REGISTRY)/ocp/4.4:oauth-proxy,./Dockerfile.rhel7,.)
+
+# Run core verification and all self contained tests.
+check: | verify test-unit
+.PHONY: check
 
 clean:
 	$(RM) ./oauth-proxy
 .PHONY: clean
+
+GO_TEST_PACKAGES :=./ ./api ./cookie ./providers
```

Comment:
is there any specific one you want to ignore? otherwise `./...` would seem more fitting

Reply:
ignoring vendor/ here

Reply:
vendor is already ignored for a few golang releases in `./...`, did modules change it?

Reply:
Hm, you're right, I thought I saw some vendor failures, whereas in fact these might have been failure that appear because of the `test/` directory which only contains the e2e tests and is the actual directory that needs to be avoided.

If there is a different way to have it excluded, I can do it for the repo.

For reference, the `go test ./...` failures:
```
# k8s.io/apimachinery/pkg/api/meta
vet: in k8s.io/apimachinery/pkg/api/meta, can't import facts for package "k8s.io/apimachinery/pkg/runtime": open $WORK/b157/vet.out: no such file or directory
# k8s.io/apimachinery/pkg/util/strategicpatch
vet: in k8s.io/apimachinery/pkg/util/strategicpatch, can't import facts for package "k8s.io/apimachinery/pkg/util/mergepatch": open $WORK/b389/vet.out: no such file or directory
# k8s.io/apimachinery/pkg/apis/meta/internalversion
vet: open $WORK/b381/vet.cfg: no such file or directory
open /tmp/go-build183862199/b380/vet.cfg: no such file or directory
# k8s.io/api/rbac/v1alpha1
vet: in k8s.io/api/rbac/v1alpha1, can't import facts for package "k8s.io/apimachinery/pkg/apis/meta/v1": open $WORK/b145/vet.out: no such file or directory
# k8s.io/api/rbac/v1beta1
vet: in k8s.io/api/rbac/v1beta1, can't import facts for package "k8s.io/apimachinery/pkg/apis/meta/v1": open $WORK/b145/vet.out: no such file or directory
# k8s.io/client-go/rest
vet: ../../../go/pkg/mod/k8s.io/client-go@v0.17.1/rest/client.go:27:2: could not import k8s.io/apimachinery/pkg/runtime (open $WORK/b157/_pkg_.a: no such file or directory)
open /tmp/go-build183862199/b275/vet.cfg: no such file or directory
open /tmp/go-build183862199/b373/vet.cfg: no such file or directory
open /tmp/go-build183862199/b272/vet.cfg: no such file or directory
# k8s.io/api/core/v1
can't create $WORK/b236/_pkg_.a: open $WORK/b236/_pkg_.a: no such file or directory
FAIL	github.com/openshift/oauth-proxy [build failed]
ok  	github.com/openshift/oauth-proxy/api	0.009s
ok  	github.com/openshift/oauth-proxy/cookie	0.009s
ok  	github.com/openshift/oauth-proxy/providers	0.018s
FAIL	github.com/openshift/oauth-proxy/providers/openshift [build failed]
FAIL	github.com/openshift/oauth-proxy/test/e2e [build failed]
?   	github.com/openshift/oauth-proxy/test/e2e/hello	[no test files]
ok  	github.com/openshift/oauth-proxy/util	0.091s
FAIL
```

Reply:
added a comment and fixed it to include subdirectories

Reply:
I don't think that is it. `./test/e2e/` although you don't want to actually run it as you'd want to split units and e2e, **should** technically work no matter what.

I think the real cause might be related to vendoring. But I don't yet know why the tests compile separetelly but fail to build in some combinations.

Even this fails:
```
$ go clean -testcache
$ go test -mod=vendor ./util/... ./providers/...
ok  	github.com/openshift/oauth-proxy/util	0.011s
open /tmp/go-build201163669/b091/providers.test: no such file or directory
FAIL	github.com/openshift/oauth-proxy/providers [build failed]
open /tmp/go-build201163669/b128/openshift.test: no such file or directory
FAIL	github.com/openshift/oauth-proxy/providers/openshift [build failed]
FAIL
```

When I tried to clear the deps and start again, it shows incompatible version used by your code:
```
o mod init
go: creating new go.mod: module github.com/openshift/oauth-proxy
 tnozicka@lenovo-t580  ~/go/src/github.com/openshift/oauth-proxy   pr-148 ●
 go mod tidy
go: finding github.com/openshift/library-go latest
go: finding k8s.io/api v0.17.2
go: downloading k8s.io/apiserver v0.17.2
go: downloading github.com/openshift/library-go v0.0.0-20200128110723-fa93bef5ede7
go: downloading k8s.io/api v0.17.2
go: finding github.com/bitly/go-simplejson v0.5.0
go: downloading github.com/bitly/go-simplejson v0.5.0
go: extracting github.com/openshift/library-go v0.0.0-20200128110723-fa93bef5ede7
go: extracting github.com/bitly/go-simplejson v0.5.0
go: extracting k8s.io/apiserver v0.17.2
go: extracting k8s.io/api v0.17.2
go: finding k8s.io/apimachinery v0.17.2
go: downloading k8s.io/apimachinery v0.17.2
go: extracting k8s.io/apimachinery v0.17.2
go: finding github.com/18F/hmacauth latest
go: finding github.com/bmizerany/assert latest
go: finding github.com/mreiferson/go-options v1.0.0
go: downloading github.com/mreiferson/go-options v1.0.0
go: extracting github.com/mreiferson/go-options v1.0.0
go: finding golang.org/x/net latest
go: finding github.com/yhat/wsutil latest
go: downloading k8s.io/component-base v0.17.2
go: extracting k8s.io/component-base v0.17.2
github.com/openshift/oauth-proxy/providers/openshift imports
	k8s.io/apiserver/pkg/server/dynamiccertificates imports
	k8s.io/client-go/tools/events: module k8s.io/client-go@latest found (v11.0.0+incompatible), but does not contain package k8s.io/client-go/tools/events
```

Reply:
so let's
/hold
while we figure this out

Reply:
/hold cancel
found a fun thing in the unit-tests of the `util/` repository, see commit `don't remove /tmp when unit-tests are run`